### PR TITLE
Moving JMT's codec code to scrypto repo (using SBOR derive)

### PR DESCRIPTION
**Pull request by jakrawcz-rdx**
A refactoring requested at https://github.com/radixdlt/babylon-node/pull/298/files#r1097250673.

### CODE REVIEW

File: radix-engine-stores/src/hash_tree/test.rs [hunk 1]
```diff
@@ -124,37 +127,82 @@ fn supports_empty_state() {
     assert_eq!(hash_v3, SPARSE_MERKLE_PLACEHOLDER_HASH);
 }
 
-// Note: this test relies on the impl details of underlying tree structure.
-// In particular, it is now assuming a single tree, keyed by hash(substate_id).
-// The asserts will need to be adjusted after introducing separate ReNode tree
-// vs Substate trees.
 #[test]
-#[ignore]
```
Comment (jakrawcz-rdx) [inline radix-engine-stores/src/hash_tree/test.rs:132]: Note from author: this test was very detailed and cool, but unfortunately also fragile (sensitive to how we do hashing of substate ids). Thank you @talekhinezh for tracking it down and disabling 👍 
I replaced the assert with a more coarse grained one (i.e. "some previous version nodes are marked as stale") and re-enabled.

File: radix-engine-stores/src/hash_tree/tree_store.rs [hunk 2]
```diff
@@ -56,80 +59,162 @@ pub trait ReadableTreeStore {
...
+    #[inline]
+    fn encode_value_kind(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        encoder.write_value_kind(Self::value_kind())
     }
 
-    fn record_stale_node(&mut self, key: &NodeKey) {
-        self.overlay.record_stale_node(key)
+    #[inline]
+    fn encode_body(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        encoder.write_byte(u8::from(*self))?;
```
Comment (dhedey) [inline radix-engine-stores/src/hash_tree/tree_store.rs:170]: Slightly better abstracted is to do `self.0.encode_body(encoder)`

File: radix-engine-stores/src/hash_tree/tree_store.rs [hunk 3]
```diff
@@ -56,80 +59,162 @@ pub trait ReadableTreeStore {
...
+        Ok(())
+    }
+}
+
+impl<D: Decoder<ScryptoCustomValueKind>> Decode<ScryptoCustomValueKind, D> for Nibble {
+    fn decode_body_with_value_kind(
+        decoder: &mut D,
+        value_kind: ValueKind<ScryptoCustomValueKind>,
+    ) -> Result<Self, DecodeError> {
+        decoder.check_preloaded_value_kind(value_kind, Self::value_kind())?;
```
Comment (dhedey) [inline radix-engine-stores/src/hash_tree/tree_store.rs:180]: Similarly, these two lines could be replaced with `Ok(Nibble(u8::decode_body_with_value_kind(decoder, value_kind)?))`

File: radix-engine-stores/src/hash_tree/tree_store.rs [hunk 4]
```diff
@@ -56,80 +59,162 @@ pub trait ReadableTreeStore {
...
+        (even, self.bytes()).encode_body(encoder)
+    }
+}
+
+impl<D: Decoder<ScryptoCustomValueKind>> Decode<ScryptoCustomValueKind, D> for NibblePath {
+    fn decode_body_with_value_kind(
+        decoder: &mut D,
+        value_kind: ValueKind<ScryptoCustomValueKind>,
+    ) -> Result<Self, DecodeError> {
+        decoder.check_preloaded_value_kind(value_kind, Self::value_kind())?;
```
Comment (dhedey) [inline radix-engine-stores/src/hash_tree/tree_store.rs:210]: You can skip this line as it's implicit in the call in the line below :)

**Review by dhedey — APPROVED** [review]
Nice 👍

**Pull request merged**